coding/python_testing.rst

```diff
@@ -4,6 +4,156 @@ Python Unit Testing
+
+A simple :mod:`unittest` example is shown below:
+
+.. literalinclude:: unit_test_snippets/unittest_basic_example.py
+   :language: python
+
+The important things to note in this example are:
+
+* If the test is being executed using :command:`python` from the command line the :py:func:`unittest.main` call performs the test discovery and executes the tests, setting exit status to non-zero if any of the tests fail.
+* Test classes are executed in the order in which they appear in the test file.
+  In this case the tests in :class:`DemoTestCase1` will be executed before those in :class:`DemoTestCase2`.
+* Test classes must, ultimately, inherit from :class:`unittest.TestCase` in order to be discovered, and tests themselves must be methods that start with ``test``.
+  All other methods and classes will be ignored by the test system but can be used by tests.
+* If a test method completes the test passes, if it throws an uncaught exception the test has failed.
+
+Using a more advanced test runner
+=================================
+
+All tests should be written such that they are runnable using `pytest`_, a policy adopted in `RFC-69`_.
+`pytest`_ provides a much richer execution and reporting environment for tests and can be used to run multiple tests files together.
+`pytest`_ test discovery is much more flexible than that provided by :mod:`unittest` but LSST test files should not take advantage of that flexibility as it can lead to inconsistency in test reports that depend on the specific test runner.
+In particular, care must be taken not to have free functions that use a ``test`` prefix or non-\ :class:`~unittest.TestCase` test classes that are named with a ``Test`` prefix in the test files.
+
+.. _RFC-69: https://jira.lsstcorp.org/browse/RFC-69
+
+`pytest`_ can run tests from more than one file in a single invocation and this can be used to verify that there is no state contaminating later tests.
+To run `pytest`_ use the :command:`py.test` executable:
+
+.. code-block:: shell
+
+   $ py.test tests/*.py
+
+and to ensure that the order of test execution does not matter it is useful to sometimes run the tests in reverse order:
+
+.. code-block:: shell
+
+   $ py.test `ls -r tests/*.py`
+
+When writing tests it is important that tests are skipped using the proper :mod:`unittest` :ref:`skipping framework <python:unittest-skipping>` rather than returning from the test early.
+:mod:`unittest` can support skipping of individual tests and entire classes using decorators, and exceptions can also be used.
+LSST code sometimes raises skip exceptions in :meth:`~unittest.TestCase.setUp` or :meth:`~unittest.TestCase.setUpClass` class methods and :mod:`unittest` provides much flexibility in how skipping can be implemented.
+It is also possible to indicate that a particular test is expected to fail, to the extent that it is reported it a test unexpectedly passes.
```

Comment:
"that it is reported it a test..."? Please reword. Perhaps " "and it is reported as an error if the test unexpected passes"